### PR TITLE
Unit test helpers

### PR DESCRIPTION
**Pull request by semseysandor**
I wanted to use `CRM_RcBase_Test_BaseHeadlessTestCase` in another extension as the parent test case for headless tests.
It turned out the classloader won't find this class when running the tests.

Inside a test method it is OK to use classes from other extensions (if they are installed in `setUpHeadless()` or `setUpBeforeClass()`) because then they get registered in the classloader.

So this class (`CRM_RcBase_Test_BaseHeadlessTestCase`) is not really useful.
This way I refactored the methods to different classes:
- `cvApi()`, `rawSqlQuery()` and `getNextAutoIncrementValue()` methods to a new `CRM_RcBase_Test_UtilsHeadless` library class as I think these methods could be useful in other tests. Also they are static now, so easier to use.
- `getNextExternalID()`, `getNextContactSequence()` and `nextSampleIndividual()` methods to the `CRM_RcBase_Api_ApiTestCase` testcase class. These methods are specific to API tests, so the API headless test classes inherit from this case.
- created a `CRM_RcBase_HeadlessTestCase` base headless testcase. All headless tests inherit from this class.

Also added some extra helper methods for unit testing in the `CRM_RcBase_Test_UtilsBasic` class. They are for accessing private or protected methods and properties. Maybe it will be useful in some tests.

Added tests for the new `Test_Utils` classes.

Some other refactorings:
- removed unneccessary 2nd parameter from `expectException()` and `expectExceptionMessage()`
- changed all `$this->assert` to `self::assert`

### CODE REVIEW

File: tests/phpunit/CRM/RcBase/Test/UtilsBasicTest.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+/**
+ * Test Utils Basic class
+ *
+ * @group unit
+ */
+class CRM_RcBase_Test_UtilsBasicTest extends TestCase
+{
+    public function testCallingRestrictedMethodThrowsException()
```
Comment (akosgarai) [inline tests/phpunit/CRM/RcBase/Test/UtilsBasicTest.php:27]: I don't get this test. Why is necessary? What is tested here? Same question for the next test.

Comment (semseysandor) [reply tests/phpunit/CRM/RcBase/Test/UtilsBasicTest.php:27]: Not really necessary. I was just curious what happens, but yes there's no need to be in the final version.
They should have stayed only a one-off test for me...

File: CRM/RcBase/Test/UtilsBasic.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+     * Gets a restricted property
+     *
+     * @param mixed $object Object with restricted property
+     * @param string $property Name of property
+     *
+     * @return mixed
+     *
+     * @throws \ReflectionException
+     */
+    public static function getProtectedProperty(&$object, string $property)
```
Comment (akosgarai) [inline CRM/RcBase/Test/UtilsBasic.php:18]: Is this function used anywhere? I can only see the testcase.

File: CRM/RcBase/Test/UtilsBasic.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * Sets a restricted property
+     *
+     * @param mixed $object Object with restricted property
+     * @param string $property Name of property
+     * @param mixed $value Value to set
+     *
+     * @throws \ReflectionException
+     */
+    public static function setProtectedProperty(&$object, string $property, $value): void
```
Comment (akosgarai) [inline CRM/RcBase/Test/UtilsBasic.php:40]: Is this function used anywhere? Only the testcase was found.

File: CRM/RcBase/Test/UtilsBasic.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * @param mixed &$object Object with restricted method
+     * @param string $method Name of method
+     * @param array|null $params Parameters to method
+     *
+     * @return mixed
+     *
+     * @throws \ReflectionException
+     */
+    public static function invokeProtectedMethod(&$object, string $method, array $params = null)
```
Comment (akosgarai) [inline CRM/RcBase/Test/UtilsBasic.php:64]: Same question as above.

File: CRM/RcBase/Test/UtilsBasic.php (hunk 2 above)
Comment (semseysandor) [reply CRM/RcBase/Test/UtilsBasic.php:18]: There is a test involving a Form_Base class, which is not submitted yet. This function is used there. This is not an essential function like can't test without it. Also Reflections are evil and tests shouldn't call invoked methods and check internal variables. Maybe when testing Base classes that could be useful.

Thats was the original reason, but I can let these methods go.
If you think they will be no use later and we should not rely on Reflection classes I'm happy to leave these methods behind.

**Review by akosgarai — APPROVED** [review]

**Pull request merged**